### PR TITLE
windows: avoid UNC paths in `run_ui_editor`

### PR DESCRIPTION
**Pull request by mlcui-corp**
~~**This PR depends on #3995. Please only review the last commit until that PR has been merged.**~~

See https://github.com/martinvonz/jj/issues/3986 for more details. This is a no-op for non-Windows.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: cli/tests/test_commit_command.rs [hunk 1]
```diff
@@ -73,6 +73,28 @@ fn test_commit_with_editor() {
...
+
+    std::fs::write(edit_script, ["dumppath path"].join("\0")).unwrap();
+    test_env.jj_cmd_ok(&workspace_path, &["commit"]);
+
+    let editor_path = std::fs::read_to_string(test_env.env_root().join("path")).unwrap();
+    // While `assert!(!editor_path.starts_with("//?/"))` could work here in most cases,
+    // it fails when it is not safe to strip the prefix, such as paths over 260 chars.
+    assert_eq!(
+        editor_path,
+        dunce::simplified(std::path::Path::new(&editor_path))
```
Comment (martinvonz) [inline cli/tests/test_commit_command.rs:92]: nit: make `std::path::Path` unqualified (it's already imported anyway). Same in the other tests, except for `test_describe_command`. Oh, maybe that's why you didn't import it here? I would still import it here (seems like Clippy might start complaining about it some day; i'm a bit surprised it doesn't already). I don't have much opinion whether to also import it in `test_describe_command` (guarded by `#[cfg(windows)]`) or to keep it local there.

File: cli/tests/test_commit_command.rs [hunk 2]
```diff
@@ -73,6 +73,28 @@ fn test_commit_with_editor() {
...
+fn test_commit_with_editor_avoids_unc() {
+    let mut test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["git", "init", "repo"]);
+    let workspace_path = test_env.env_root().join("repo");
+    let edit_script = test_env.set_up_fake_editor();
+
+    std::fs::write(edit_script, ["dumppath path"].join("\0")).unwrap();
+    test_env.jj_cmd_ok(&workspace_path, &["commit"]);
+
+    let editor_path = std::fs::read_to_string(test_env.env_root().join("path")).unwrap();
```
Comment (martinvonz) [inline cli/tests/test_commit_command.rs:87]: nit: is it simpler to wrap this in `Path::new()`? then you don't need to call `.to_str().unwrap()` below, i think. Same in the other tests

**Review by martinvonz — APPROVED** [review]
Thanks for working on this. I don't think we have many devs on Windows who would know what to do here (I certainly don't).

File: cli/tests/test_commit_command.rs (hunk 2 above)
Comment (mlcui-corp) [reply cli/tests/test_commit_command.rs]: Done - also applied in #3995.

File: cli/tests/test_commit_command.rs (hunk 1 above)
Comment (mlcui-corp) [reply cli/tests/test_commit_command.rs]: I've imported it inside these just in case we remove the other `std::path::Path` import. PTAL

**Comment by mlcui-corp** [comment]
Resolved comments - PTAL.

File: cli/tests/test_commit_command.rs [hunk 3]
```diff
@@ -73,6 +73,27 @@ fn test_commit_with_editor() {
     "###);
 }
 
+#[test]
+#[cfg(windows)]
```
Comment (yuja) [inline cli/tests/test_commit_command.rs:77]: nit: If the test passes on the other platforms, I would remove `cfg(windows)` (and move `use std::path::PathBuf` to top.) Windows-only code paths are easier to break.

**Review by yuja — APPROVED** [review]

File: cli/tests/test_commit_command.rs (hunk 3 above)
Comment (mlcui-corp) [reply cli/tests/test_commit_command.rs]: Done.

File: cli/tests/test_commit_command.rs (hunk 1 above)
Comment (mlcui-corp) [reply cli/tests/test_commit_command.rs]: Resolved as part of the parent PR.

**Pull request merged**